### ISSUE TITLE
Provide a custom implementation of Optional for use with BindsOptionalOf

### ISSUE THREAD

**Issue by Blisse** [google/dagger#1539]
BindsOptionalOf is only compatible with the guava Optional and the java Optional. On Android, these are somewhat non-starters. guava is widely recognized as too large of a dependency for mobile, and java's Optional isn't available until min 24, which just isn't feasible.

It would be preferable if Dagger could provide a third DaggerOptional class that could be used with BindsOptionalOf as well, to unlock that feature for Android.

**Comment by JakeWharton** [comment]
The D8/R8 team is working on desugaring Optional. Star https://issuetracker.google.com/issues/114481425 for updates. If you peek in the R8 repo, you might just spot the fact that it's already partially implemented and should be available in an alpha release of the Android Gradle plugin soon. As a result, I'm against Dagger doing this.

**Comment by Blisse** [comment]
Thanks Jake! Good to hear this is on some roadmap.

Would you happen to have an ETA for when Android developers would be able to take advantage of these additional de-sugarings? If it's within a month or so then that's fair, but if the timeline is more like 1-2 years then I'd feel much more strongly about being unblocked here.

**Comment by JakeWharton** [comment]
I would estimate that it would be a feature of 3.7 which I would guess are 1-2 months away from being available as alpha. It's not guaranteed to be in the first alpha, but it would hopefully show up in an early alpha for testing.

**Comment by bcorso** [comment]
Closing this since it looks like https://issuetracker.google.com/issues/114481425 is fixed and the requested feature is no longer needed from Dagger.
